Fix project_settings entry creation for global scripts.

ericrybick opened the pull request:
The global script entries in project settings were created outside the global section. This caused the `Property not found: global/_global_script_class_icons` warning when clicking on Global project setting.

I've added some code to remove incorrect global script entries for existing projects because the properties would appear twice in the global section without removing them.

fixes #26247 

### Issue Visualization:
![wqgqwfqw](https://user-images.githubusercontent.com/8831226/53465788-a2394380-3a4f-11e9-86e8-3a42b552f13e.png)

### Without incorrect global script entry removal
![4g34h46j56j](https://user-images.githubusercontent.com/8831226/53465865-e75d7580-3a4f-11e9-8541-33d58ce98546.PNG)

akien-mga commented on core/project_settings.cpp:
```diff
@@ -284,6 +284,17 @@ void ProjectSettings::_convert_to_last_version(int p_from_version) {
 		}
 	}
+
+	// Deletes invalid global script entries from project settings.
```
I don't think this is needed, `editor/editor_sectioned_inspector.cpp` already filters out all properties starting with `global/`.

akien-mga commented on core/script_language.cpp:
```diff
@@ -214,7 +214,7 @@ void ScriptServer::save_global_classes() {
 	}
 
-	ProjectSettings::get_singleton()->set("_global_script_classes", gcarr);
+	ProjectSettings::get_singleton()->set("global/_global_script_classes", gcarr);
```
You could make it `global/script_classes` and `global/script_class_icons`.

ericrybick replied:
Do you think that is necessary? Since `editor/editor_data.cpp` already sets `global/script_class_icons`.

akien-mga replied:
I mean this one should be `global/script_classes`, and the one in `editor_data.cpp` should be `global/script_class_icons`. There's no reason for those two settings to be without a section (the original bug).

ericrybick replied:
I should read more thoroughly. Gotcha!

ericrybick commented:
@akien-mga 

Ok after a few checks, I found that the filter in `editor/editor_sectioned_inspector.cpp` caused the problem. I have changed my commit and removed the cleaning part. `project.godot` will keep the wrong entries now (if the project has been created before this fix), but they won't be used by/for anything.

akien-mga commented on editor/editor_sectioned_inspector.cpp:
```diff
@@ -78,15 +78,10 @@ class SectionedInspectorFilter : public Object {
 				continue;
 
-			if (sp == -1) {
-				pi.name = "global/" + pi.name;
```
Removing this might break some stuff, I wouldn't merge that for 3.1 at any rate.

ericrybick replied:
How? All this does is adding global/ to all project settings without a section.

ericrybick replied:
akien expained it to me:
Normally the project settings are supposed to hide/ignore anything that starts with "global/", which is why the sectioned property editor adds it to settings without section.

akien-mga commented:
I'm going to give a try to another approach that we discussed with @reduz a few days ago:
```
[14:26:58] <Akien> reduz: Before we release 3.1, I'd also like to move those two lines to a proper section:
[14:26:59] <Akien> _global_script_classes=null
[14:26:59] <Akien> _global_script_class_icons=null
[14:27:08] <Akien> (in project.godot)
[14:27:28] <Akien> It's weird to have those at the top as if they were more important than app name and whatnot
[14:27:57] <Akien> What section should I use for those? [global], or [meta]?
[14:28:49] <reduz> maybe the properties could be
[14:29:10] <reduz> scripting/_global_script_classes
[14:29:21] <reduz> i think the project settings editor will hide them anyway
[14:30:09] <Akien> Currently they're shown in the Project Settings, in a bogus way
[14:30:18] <Akien> https://pasteboard.co/I1AdnLK.png
[14:30:34] <Akien> Opening that "Global" section triggers
[14:30:34] <Akien> WARNING: _get: Property not found: global/_global_script_classes
[14:30:34] <Akien>    At: core/project_settings.cpp:193.
[14:30:34] <Akien> WARNING: _get: Property not found: global/_global_script_class_icons
[14:30:34] <Akien>    At: core/project_settings.cpp:193.
[14:30:41] <reduz> ah that sucks they should be hidden
[14:30:45] <Akien> Yeah
[14:30:54] <Akien> I'll have a look at that. [scripting] sounds goodd
[14:30:56] <Akien> *good
[14:30:59] <reduz> actually i think we should probably save this info different
[14:31:00] <reduz> like
[14:31:07] <reduz> [global_scripts_classes]
[14:31:13] <reduz> name=[path,icon]
[14:31:18] <reduz> this way its more VCS friendly
[14:31:42] <reduz> if two different users add one of these, there will be no conflict
[14:31:43] <Akien> Currently it saves info this way: https://hastebin.com/ehisofilew.makefile
[14:31:57] <reduz> since class names can be a property name anyway
```

akien-mga commented:
I've tried to implement the solution mentioned in the previous comment with:
```diff
diff --git a/core/script_language.cpp b/core/script_language.cpp
index f0310ffc3..d95922bcb 100644
--- a/core/script_language.cpp
+++ b/core/script_language.cpp
@@ -108,14 +108,14 @@ void ScriptServer::init_languages() {
 
 	{ //load global classes
 		global_classes_clear();
-		if (ProjectSettings::get_singleton()->has_setting("_global_script_classes")) {
-			Array script_classes = ProjectSettings::get_singleton()->get("_global_script_classes");
-
+		if (ProjectSettings::get_singleton()->has_section("global_script_classes")) {
+			PoolStringArray script_classes = ProjectSettings::get_singleton()->get_section_keys("global_script_classes");
 			for (int i = 0; i < script_classes.size(); i++) {
-				Dictionary c = script_classes[i];
-				if (!c.has("class") || !c.has("language") || !c.has("path") || !c.has("base"))
+				String class_name = script_classes[i];
+				Dictionary d = ProjectSettings::get_singleton()->get_value("global_script_classes", class_name);
+				if (!d.has("language") || !d.has("path") || !d.has("base"))
 					continue;
-				add_global_class(c["class"], c["base"], c["language"], c["path"]);
+				add_global_class(class_name, d["base"], d["language"], d["path"]);
 			}
 		}
 	}
@@ -204,17 +204,15 @@ void ScriptServer::get_global_class_list(List<StringName> *r_global_classes) {
 void ScriptServer::save_global_classes() {
 	List<StringName> gc;
 	get_global_class_list(&gc);
-	Array gcarr;
 	for (List<StringName>::Element *E = gc.front(); E; E = E->next()) {
 		Dictionary d;
-		d["class"] = E->get();
-		d["language"] = global_classes[E->get()].language;
-		d["path"] = global_classes[E->get()].path;
-		d["base"] = global_classes[E->get()].base;
-		gcarr.push_back(d);
+		String class_name = E->get();
+		d["language"] = global_classes[class_name].language;
+		d["path"] = global_classes[class_name].path;
+		d["base"] = global_classes[class_name].base;
+		ProjectSettings::get_singleton()->set("global_script_classes/" + class_name, d);
 	}
 
-	ProjectSettings::get_singleton()->set("_global_script_classes", gcarr);
 	ProjectSettings::get_singleton()->save();
 }
 
diff --git a/editor/editor_data.cpp b/editor/editor_data.cpp
index b87b80daf..f0c1bc60c 100644
--- a/editor/editor_data.cpp
+++ b/editor/editor_data.cpp
@@ -940,26 +940,24 @@ void EditorData::script_class_save_icon_paths() {
 
 	Dictionary d;
 	for (List<StringName>::Element *E = keys.front(); E; E = E->next()) {
-		if (ScriptServer::is_global_class(E->get()))
-			d[E->get()] = _script_class_icon_paths[E->get()];
+		if (ScriptServer::is_global_class(E->get())) {
+			ProjectSettings::get_singleton()->set("global_script_class/" + E->get(), _script_class_icon_paths[E->get()]);
+		}
 	}
 
-	ProjectSettings::get_singleton()->set("_global_script_class_icons", d);
 	ProjectSettings::get_singleton()->save();
 }
 
 void EditorData::script_class_load_icon_paths() {
 	script_class_clear_icon_paths();
 
-	if (ProjectSettings::get_singleton()->has_setting("_global_script_class_icons")) {
-		Dictionary d = ProjectSettings::get_singleton()->get("_global_script_class_icons");
-		List<Variant> keys;
-		d.get_key_list(&keys);
-
-		for (List<Variant>::Element *E = keys.front(); E; E = E->next()) {
-			String name = E->get().operator String();
-			_script_class_icon_paths[name] = d[name];
-
+	if (ProjectSettings::get_singleton()->has_section("global_script_classes")) {
+		PoolStringArray script_classes = ProjectSettings::get_singleton()->get_section_keys("global_script_classes");
+		for (int i = 0; i < script_classes.size(); i++) {
+			String class_name = script_classes[i];
+			Dictionary d = ProjectSettings::get_singleton()->get_value("global_script_classes", class_name);
+			String icon_path = d.get("icon", "");
+			_script_class_icon_paths[class_name] = icon_path;
 			String path = ScriptServer::get_global_class_path(name);
 			script_class_set_name(path, name);
 		}
```

The only issue is that I assumed `ProjectSettings` offered the same public API as `ConfigFile` (`has_section`, `get_section_keys`), but it doesn't :/ So not sure how to properly implement it.

ericrybick committed:
Hide global section in project settings when no property is defined
Prevent _global_script entries in project settings without section from being loaded

fixes #26247

akien-mga commented:
Thanks!
